text/3668-async-closure.md

```diff
@@ -0,0 +1,625 @@
+```
+
+### Interaction with return-type notation, naming the future returned by calling
+
+With `async Fn() -> T` trait bounds, we don't know anything about the `Future` returned by calling the async closure other than that it's a `Future` and awaiting that future returns `T`.
+
+This is not always sufficient, for example, if you want to spawn a future onto another thread:
+
+```rust
+async fn foo(x: impl async Fn(&str)) -> Result<()> {
+    tokio::spawn(x("hello, world")).await
+}
+```
+
+```
+error[E0277]: cannot be sent between threads safely
+   --> src/lib.rs
+    |
+    |     tokio::spawn(x("hello, world")).await
+    |     ------------ ^^^^^^^^^^^^^^^^^ cannot be sent between threads safely
+    |     |
+    |     required by a bound introduced by this call
+```
+
+With the acceptance of the RTN (return-type notation) [RFC 3654](https://github.com/rust-lang/rfcs/pull/3654), this RFC specifies that users will be allowed to add RTN-like bounds to type parameters that are also bounded by `async Fn()`. Concretely, this bound expands to bound both `CallOnceFuture` and `CallRefFuture` (if the latter exists):
+
+```rust
+async fn foo(x: F) -> Result<()>
+where
+    F: async Fn(&str) -> Result<()>,
+    // The future from calling `F` is `Send` and `'static`.
+    F(..): Send + 'static,
+    // Which expands to two bounds:
+    // `for<'a> <F as AsyncFnMut>::CallRefFuture<'a>: Send`
+    // `<F as AsyncFnOnce>::CallOnceFuture: Send`
+    // the latter is only if `F` is bounded with `async Fn` or `async FnMut`.
+{
+    tokio::spawn(x("hello, world")).await
+}
+```
+
+This bound is only valid if there is a corresponding `async Fn*()` trait bound.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Why do we need a new set of `AsyncFn*` traits?
+
+As demonstrated in the motivation section, we need a set of traits that are *lending* in order to represent futures which borrow from the closure's captures. We technically only need to add `LendingFn` and `LendingFnMut` to our lattice of `Fn*` traits, leaving us with a hierarchy of traits like so:
+
+```mermaid
+flowchart LR
+
+Fn
+FnMut
+FnOnce
+LendingFn
+LendingFnMut
+
+Fn -- isa --> FnMut
+FnMut -- isa --> FnOnce
+
+LendingFn -- isa --> LendingFnMut
+
+Fn -- isa --> LendingFn
+FnMut -- isa --> LendingFnMut
+```
+
+However, there are some concrete technical implementation details that limit our ability to use `LendingFn` ergonomically in the compiler today. These have to do with:
+
+- Closure signature inference.
+- Limitations around higher-ranked trait bounds.
+- Shortcomings with error messages.
+
+These limitations, plus the fact that the underlying trait should have no effect on the user experience of async closures and async `Fn` trait bounds, leads us to `AsyncFn*` for now. To ensure we can eventually move to these more general traits, we reserved the precise `AsyncFn*` trait definitions (including the associated types) as an implementation detail.
```

Comment:
I'm honestly really confused why not giving a stable to `Output` is such a problem here. Can you point me to code that actually uses `FnOnce::Output` anywhere? 

One unique thing about `Fn` trait goals is that their output is *necessarily* constrained by an associated type bound. You can never write:

```
fn foo<F: FnOnce<()>>(f: F) {
    let ret: /* the only name for this type is `F::Output` */ = f();
}
```

Which is the only case where the output must be named by `T::Output`, since it's an unconstrained (or in type system parlance, "rigid") associated type. You must always write:

```
fn foo<F: FnOnce() -> T>(f: F) {
    let ret: /* T */ = f();
}
```

In which case, the output type is necessarily named `T`. While it's possible that you may write `F::Output` in the latter example, I find that to be a curiosity and not a serious thing that people write commonly.

Reply:
The only case where I can see this being necessary is a macro that cannot name the return type because of some token shenanigans or because of lifetime names and needs to be able to name it via `F::Output`, but even then I'd need to see something specific. I'm personally totally happy with maintaining this inconsistency for the time being.